### PR TITLE
Fix context end calls

### LINKED ISSUE #1710: Cannot retrieve generated keys with SQLInsertClause::executeWithKeys if SQLCloseListener is enabled
Hi. Suppose I have the following setup:

    SQLQueryFactory qf = new SQLQueryFactory(mySchemaConfig, myDataSource, true); <-- add sql close listener
    SQLInsertClause todoInsert = qf.insert(myTodoQType);
    Todo myTodoPojo = new Todo(); // Todo table has ID and text columns.

And then I try to insert a TODO item into my db, and retrieve it's ID, based on the next sequence number, preconfigured in my database schema:

    myTodo.setText("Buy groceries");
    Long insertedTodoId = todoInsert
        .set(myTodoQType.id, SQLExpressions.nextVal("MY_TODO_ID_SEQUENCE"))
        .set(myTodoQType.text, myTodo.getText())
        .executeWithKey(myTodoQType.id);
    myTodo.setId(insertedTodoId);

In the code above, I'm finding that if I configure my query factory `qf` to use the `SQLCloseListener`, it will cause this line:

    Long insertedTodoId = todoInsert
        .set(myTodoQType.id, SQLExpressions.nextVal("MY_TODO_ID_SEQUENCE"))
        .set(myTodoQType.text, myTodo.getText())
        .executeWithKey(myTodoQType.id);

to throw this exception:

    com.querydsl.core.QueryException: java.sql.SQLException: Closed Resultset: next
	at com.querydsl.sql.DefaultSQLExceptionTranslator.translate(DefaultSQLExceptionTranslator.java:39)
	at com.querydsl.sql.Configuration.translate(Configuration.java:441)
	at com.querydsl.sql.dml.SQLInsertClause.executeWithKey(SQLInsertClause.java:179)
	at com.querydsl.sql.dml.SQLInsertClause.executeWithKey(SQLInsertClause.java:153)

Now, if I reconfigure my query factory `qf` to not use the `SQLCloseListener`:

    SQLQueryFactory qf = new SQLQueryFactory(mySchemaConfig, myDataSource, false);

The `insert` clause will run fine, but my code starts leaking connections if I don't find a way to close them explicitly.

1) Is my understanding on how to use `executeWithKey` correct?
2) Should I manage my own connections with `insert` clauses instead of relying on `SQLCloseListener`? Or is this a bug in the listener itself?

Thanks for your time and help!

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1710

### CODE REVIEW

**Comment by Shredder121** [comment]
This looks good to me.

File: querydsl-sql/src/main/java/com/querydsl/sql/SQLListeners.java [hunk 1]
```diff
@@ -38,7 +39,7 @@
     @Nullable
     private final SQLDetailedListener parent;
 
-    private final List<SQLDetailedListener> listeners = Lists.newArrayList();
+    private final Set<SQLDetailedListener> listeners = new LinkedHashSet<SQLDetailedListener>();
```
Comment (johnktims) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLListeners.java]: You could use `Sets.newLinkedHashSet()`.

**Comment by Shredder121** [comment]
@johnktims So it's good to merge now?

**Comment by johnktims** [comment]
Yes.

**Pull request merged**